explore/marginal-gain/1.Marginal-Gain-Multiple-Mutations.py

```diff
@@ -0,0 +1,253 @@
+
+
+# In[4]:
+
+get_ipython().run_cell_magic('time', '', "path = os.path.join('..', '..', 'download', 'expression-matrix.tsv.bz2')\nexpression = pd.read_table(path, index_col=0)")
+
+
+# In[5]:
+
+get_ipython().run_cell_magic('time', '', "path = os.path.join('..','..','download', 'mutation-matrix.tsv.bz2')\nY = pd.read_table(path, index_col=0)")
+
+
+# In[6]:
+
+X = {}
+X['model a'] = covariates
+X['model b'] = pd.concat([covariates,expression], axis=1)
+for k in ['model a','model b']:
+    print(X[k].shape)
+
+
+# In[80]:
+
+mutations = {
+    '7157': 'TP53',   # tumor protein p53
+    '7428': 'VHL',    # von Hippel-Lindau tumor suppressor
+    '29126': 'CD274', # CD274 molecule
+    '672': 'BRCA1',   # BRCA1, DNA repair associated
+    '675': 'BRCA2',   # BRCA2, DNA repair associated
+    '238': 'ALK',     # anaplastic lymphoma receptor tyrosine kinase
+    '4221': 'MEN1',   # menin 1
+    '5979': 'RET',    # ret proto-oncogene
+}
+
+
+# ## Median absolute deviation feature selection
+
+# In[81]:
+
+def fs_mad(x, y):
+    """    
+    Get the median absolute deviation (MAD) for each column of x
+    """
+    scores = mad(x) 
+    return scores, np.array([np.NaN]*len(scores))
+
+
+# ## Define pipeline and Cross validation model fitting
+
+# In[82]:
+
+# Parameter Sweep for Hyperparameters
+
+param_grid = {
+    'classify__loss': ['log'],
+    'classify__penalty': ['elasticnet'],
+    'classify__alpha': [1],#10.0 ** np.linspace(-3, 1, 10),
+    'classify__l1_ratio': [0.15],
+}
+
+expression_feats = Pipeline(steps=[
+    ('dim_red', FunctionTransformer(lambda X: X[:,covariates.shape[1]:])),
+    ('select', SelectKBest(fs_mad,2000)),
+])
+
+covariate_feats = Pipeline(steps=[
+    ('dim_red', FunctionTransformer(lambda X: X[:,:covariates.shape[1]])),
+])
+
+combo_pipeline = Pipeline([
+    ('imputer', Imputer()),
+    ('standardize', StandardScaler()),
+    ('features', FeatureUnion([
+        ('covariates_feats', covariate_feats),           
+        ('expression_feats', expression_feats)
+    ])),
+    ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
+])
+
+covariates_pipeline = Pipeline(steps=[
+    ('imputer', Imputer()),
+    ('standardize', StandardScaler()),
+    ('select', SelectKBest(fs_mad,'all')),
+    ('classify', SGDClassifier(random_state=0, class_weight='balanced'))
+])
+
+cv_pipeline = {}
+cv_pipeline['model a'] = GridSearchCV(estimator=covariates_pipeline, param_grid=param_grid, scoring='roc_auc')
+cv_pipeline['model b'] = GridSearchCV(estimator=combo_pipeline, param_grid=param_grid, scoring='roc_auc')
+
+
+# ## Functions to get statistics for a given model 
+
+# In[83]:
+
+# Get statistics for a given model. 
+
+def get_aurocs(X, y, pipeline, series, model_type):
+    """
+    Fit the classifier for the given mutation (y) and output predictions for it
+    """
+    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0)
+    y_train[0] = 1
```

Comment:
> (Also, note that to make this run, I have set the 1st value of the y_test and y_train to 1 just to ensure at least one positive mutation example. Not sure--could be affecting the results a bit.).

@joshlevy89, I'm concerned about this workaround. I think specifying the `stratify=y` in [`train_test_split`](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) should resolve this problem. If not, I don't think we want to make a workaround -- the failure shouldn't be masked.